terminal/runereader_posix.go

```diff
@@ -76,36 +81,126 @@ func (rr *RuneReader) ReadRune() (rune, int, error) {
 		if err != nil {
 			return r, size, err
 		}
-		if r != '[' {
-			return r, size, fmt.Errorf("Unexpected Escape Sequence: %q", []rune{'\033', r})
-		}
-		r, size, err = rr.state.reader.ReadRune()
-		if err != nil {
-			return r, size, err
-		}
+
```

Comment:
What I had in mind wasn't quite so verbose---only one of the switches varies based on O vs [.
Also, a single doc link is enough. Readers can search the docs if they need to.

How about something more like this:

```
func (rr *RuneReader) ReadRune() (rune, int, error) {
	r, size, err := rr.state.reader.ReadRune()
	if err != nil {
		return r, size, err
	}

	// ANSI terminal controls.
	// See https://vt100.net/docs/vt102-ug/appendixc.html
	const (
		escape            = '\x1b'
		normalKeypad      = '['
		applicationKeypad = 'O'
	)

	// Parse escape sequences such as ESC [ A for arrow keys.
	if r == escape {
		if rr.state.reader.Buffered() == 0 {
			// no more characters so must be Escape key
			return KeyEscape, 1, nil
		}
		r, size, err = rr.state.reader.ReadRune()
		if err != nil {
			return r, size, err
		}

		// ESC O ... or ESC A ...?
		var normal bool
		switch r {
		case normalKeypad:
			normal = true

		case applicationKeypad:
			normal = false

		default:
			return r, size, fmt.Errorf("unexpected escape sequence from terminal: %q", []rune{'\033', r})
		}

		r, size, err := rr.state.reader.ReadRune()
		if err != nil {
			return r, size, err
		}

		switch r {
		case 'A': // ESC [ A or ESC O A
			return KeyArrowUp, 1, nil
		case 'B': // ESC [ B or ESC O B
			return KeyArrowDown, 1, nil
		case 'C': // ESC [ C or ESC O C
			return KeyArrowRight, 1, nil
		case 'D': // ESC [ D or ESC O D
			return KeyArrowLeft, 1, nil
		case 'F': // ESC [ F or ESC O F
			return SpecialKeyEnd, 1, nil
		case 'H': // ESC [ H or ESC O H
			return SpecialKeyHome, 1, nil
		case '3': // ESC [ 3
			if normal {
				// discard the following '~' key from buffer
				_, _ = rr.state.reader.Discard(1)
				return SpecialKeyDelete, 1, nil
			}
		}

		// discard the following '~' key from buffer
		_, _ = rr.state.reader.Discard(1)
		return IgnoreKey, 1, nil
	}
}
```

Reply:
sorry but you are saying the opposite of @mislav, I will apply your suggestions but it's a bit difficult for me to don't have a unique voice or a maintainer of this project on this topic.

Reply:
Sorry for not being more precise earlier. BTW, neither @mislav nor I are maintainers of this repo, so you don't have to listen to anything we say. :) @AlecAivazis is the authority in these parts.